src/scripts/main.js

```diff
@@ -1,3 +1,237 @@
+      i++;
+    }
+  }
+
+  setRandom();
+  setRandom();
+  document.addEventListener('keyup', handler);
+}
+
+function hasEmptyTile() {
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      if (board[r][c] === 0) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+function setRandom() {
+  if (!hasEmptyTile()) {
+    return;
+  }
+
+  let isFound = false;
+
+  while (!isFound) {
+    const r = Math.floor(Math.random() * rows);
+    const c = Math.floor(Math.random() * columns);
+
+    if (board[r][c] === 0) {
+      board[r][c] = Math.random() < 0.1 ? 4 : 2;
+
+      const tile = document.getElementById(`${r}-${c}`);
+
+      tile.innerText = '2';
+      tile.classList.add('field-cell--2');
+      isFound = true;
+    }
+  }
+}
+
+function updateTile(tile, num) {
+  tile.innerText = '';
+  tile.classList.value = '';
+  tile.classList.add('field-cell');
+
+  if (num > 0) {
+    tile.innerText = num;
+    tile.classList.add(`field-cell--${num}`);
+  }
+}
+
+const handler = (e) => {
+  switch (e.code) {
+    case 'ArrowLeft': {
+      slideLeftAndRight(true);
+      setRandom();
+
+      break;
+    }
+
+    case 'ArrowRight': {
+      slideLeftAndRight(false);
+      setRandom();
+
+      break;
+    }
+
+    case 'ArrowUp': {
+      slideUpAndDown(true);
+      setRandom();
+
+      break;
+    }
+
+    case 'ArrowDown': {
+      slideUpAndDown(false);
+      setRandom();
+
+      break;
+    }
+  }
+
+  lose();
+  win();
+};
+
+function filterZero(row) {
+  return row.filter(num => num !== 0);
+}
+
+function slide(row) {
+  let newRow = filterZero(row);
+
+  for (let i = 0; i < newRow.length; i++) {
+    if (newRow[i] === newRow[i + 1]) {
+      newRow[i] *= 2;
+      newRow[i + 1] = 0;
+      score.textContent = `${+score.textContent + newRow[i]}`;
+    }
+  }
+
+  newRow = filterZero(newRow);
+
+  while (newRow.length < columns) {
+    newRow.push(0);
+  }
+
+  return newRow;
+}
+
+function slideLeftAndRight(left) {
+  for (let i = 0; i < rows; i++) {
+    let row = board[i];
+
+    if (left) {
+      row = slide(row);
+      board[i] = row;
+    } else {
+      row.reverse();
+      row = slide(row);
+      row.reverse();
+      board[i] = row;
+    }
+
+    for (let c = 0; c < columns; c++) {
+      const tile = document.getElementById(`${i}-${c}`);
+      const num = board[i][c];
+
+      updateTile(tile, num);
+    }
+  }
+}
+
+const slideUpAndDown = (up) => {
+  for (let c = 0; c < columns; c++) {
+    let row = [board[0][c],
+      board[1][c],
+      board[2][c],
+      board[3][c]];
+
+    if (up) {
+      row = slide(row);
+    } else {
+      row.reverse();
+      row = slide(row);
+      row.reverse();
+    }
+
+    for (let r = 0; r < columns; r++) {
+      board[r][c] = row[r];
+
+      const tile = document.getElementById(`${r}-${c}`);
+      const num = board[r][c];
+
+      updateTile(tile, num);
+    }
+  }
+};
+
+function lose() {
+  if (hasEmptyTile()) {
+    return;
+  }
```

Comment:
This check better use outside from function. From your function name is not obvious what it does. It really has conditions and if all is well, it checks the value and makes some text visible but not lose